### PR TITLE
Greatly expand style guidelines, enforce linting rules

### LINKED ISSUE #500: Expand and enforce linting guidelines
Let's talk about linting.  As noted in the docs, there isn't really a consistent style applied to Twinkle.  That's okay!  But with around 17,000 lines of code, maybe it's also okay to have some more options in place.  As of #400 we have the basics via `eslint:recommended`, and `no-caller` was added in cd45282 (`max-len` is listed but unused).

I went through the [eslint rules](https://eslint.org/docs/rules/) and picked out a bunch of possibilities; the list is a combination of options that we already follow, options we don't, options I just personally prefer, and options that I thought might seem reasonable to at least someone out there.  Some are listed with different configurations.  This is a _long_ list; still, it is by no means meant to be exhaustive or limiting, so feel free to proffer others.  Many are probably unnecessary or overkill.

I generated a table with the current results from each, arranged alphabetically by category, and intend to update the ✔️ or ❌ marks as this conversation goes forward.  The links will take you to the online docs page for each option, which should explain each in full.  I've made some (heavy-handed?) suggestions for a lot of them, in particular those with multiple configurations listed, to get us started, but feel free to disagree with those choices.

This isn't particularly urgent (see also #477), and there are a number of PRs we should review/merge beforehand to avoid conflicts, but once we've got a firm list I'll go through and make the changes.

cc @MusikAnimal @atlight @kevinji @azatoth

| Use?      | Option      | Config      | Curr errors | Fix?    | Category    | Notes?      |
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| ❌  | [no-extra-parens](https://eslint.org/docs/rules/no-extra-parens) | "error" | 146 | Yes | Possible errors | Alt|
| ❌  |   [no-extra-parens](https://eslint.org/docs/rules/no-extra-parens) | ["error", "functions"] | 0 | Yes | Possible errors | Alt; might be allow additional clarity in ternaries|
| ✔️  |   [no-extra-parens](https://eslint.org/docs/rules/no-extra-parens) | ["error", "all", { "nestedBinaryExpressions": false }] | 113 | Yes | Possible errors | Alt; required for `no-mixed-operators` |
| ✔️  |   [no-prototype-builtins](https://eslint.org/docs/rules/no-prototype-builtins) | "error" | 1 | No |  Possible errors| |
| ✔️  |   [block-scoped-var](https://eslint.org/docs/rules/block-scoped-var) | "error" | 0 | No | Best practices | |
| ✔️  |   [curly](https://eslint.org/docs/rules/curly) | "error" | 20 | Yes | Best practices | Alt|
| ❌  |   [curly](https://eslint.org/docs/rules/curly) | ["error", "multi"] | 787 | Yes | Best practices | Alt|
| ❌  |   [curly](https://eslint.org/docs/rules/curly) | ["error", "multi-line"] | 5 | Yes | Best practices | Alt|
| ✔️  |   [default-case](https://eslint.org/docs/rules/default-case) | "error" | 2 | No | Best practices | |
| ✔️  |   [eqeqeq](https://eslint.org/docs/rules/eqeqeq) | "error" | 8 | Mixed | Best practices | Alt|
| ❌  |   [eqeqeq](https://eslint.org/docs/rules/eqeqeq) | ["error", "smart"] | 8 | Mixed | Best practices | Alt|
| ✔️  |   [guard-for-in](https://eslint.org/docs/rules/guard-for-in) | "error" | 0 | No | Best practices | |
| ✔️  |   [no-else-return](https://eslint.org/docs/rules/no-else-return) | "error" | 6 | Yes | Best practices | |
| ❌  |   [no-empty-function](https://eslint.org/docs/rules/no-empty-function) | "error" | 2 | No | Best practices | |
| ❌  |   [no-eq-null](https://eslint.org/docs/rules/no-eq-null) | "error" | 0 | No |  Best practices| |
| ❔  |   [no-global-assign](https://eslint.org/docs/rules/no-global-assign) | "error" | 0 | No | Best practices | See also `no-undefined` and `no-shadow-restricted-names`|
| ✔️  |   [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion) | ["error", { "boolean": false }] | 0 | Yes | Best practices | Alt|
| ❌  |   [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion) | ["error"] | 18 | Yes | Best practices | Alt|
| ✔️  |   [no-lone-blocks](https://eslint.org/docs/rules/no-lone-blocks) | "error" | 0 | No | Best practices | |
| ✔️  |   [no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces) | ["error", { "ignoreEOLComments": true }] | 17 | Yes | Best practices | Alt|
| ❌  |   [no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces) | ["error"] | 251 | Yes | Best practices | Alt|
| ❔  |   [no-script-url](https://eslint.org/docs/rules/no-script-url) | "error" | 3 | No | Best practices | |
| ✔️  |   [no-useless-catch](https://eslint.org/docs/rules/no-useless-catch) | "error" | 0 | No | Best practices | |
| ✔️  |   [no-useless-return](https://eslint.org/docs/rules/no-useless-return) | "error" | 0 | Yes | Best practices | |
| ✔️  |   [yoda](https://eslint.org/docs/rules/yoda) | "error" | 0 | Yes | Best practices | |
| ❔  |   [no-shadow-restricted-names](https://eslint.org/docs/rules/no-shadow-restricted-names) | "error" | 2 | No | Variables | Dislikes the format of the anonymous functions in `twinkle.js` and `morebits.js`; see also `no-undefined`|
| ❔  |   [no-undefined](https://eslint.org/docs/rules/no-undefined) | "error" | 24 | No | Variables | Alternatively, can use `no-global-assign` and `no-shadow-restricted-names` to allowed usage of `undefined` while still being safe, but either way, they dislike the format of the anonymous functions in `twinkle.js` and `morebits.js`|
| ❔  |   [array-bracket-spacing](https://eslint.org/docs/rules/array-bracket-spacing) | ["error", "never"] | 127 | Yes | Stylistic | |
| ❔  |   [array-bracket-spacing](https://eslint.org/docs/rules/array-bracket-spacing) | ["error", "never", { "objectsInArrays": true }] | 126 | Yes | Stylistic | |
| ❔  |   [array-bracket-spacing](https://eslint.org/docs/rules/array-bracket-spacing) | ["error", "always"] | 52 | Yes | Stylistic | |
| ✔️  |   [block-spacing](https://eslint.org/docs/rules/block-spacing) | "error" | 0 | Yes | Stylistic | |
| ❌  |   [brace-style](https://eslint.org/docs/rules/brace-style) | ["error", "1tbs", { "allowSingleLine": true }] | 33 | Yes (27) | Stylistics | Alt|
| ✔️  |   [brace-style](https://eslint.org/docs/rules/brace-style) | ["error", "1tbs"] | 75 | Yes (69) | Stylistic | Alt|
| ✔️  |   [comma-dangle](https://eslint.org/docs/rules/comma-dangle) | "error" | 5 | Yes | Stylistic | |
| ❌  |   [comma-dangle](https://eslint.org/docs/rules/comma-dangle) | ["error", "always-multiline"] | 1326 | Yes | Stylistic | |
| ✔️  |   [comma-spacing](https://eslint.org/docs/rules/comma-spacing) | ["error", { "before": false, "after": true }] | 22 | Yes | Stylistic | |
| ✔️  |   [computed-property-spacing](https://eslint.org/docs/rules/computed-property-spacing) | ["error", "never"] | 42 | Yes | Stylistic | |
| ✔️  |   [func-call-spacing](https://eslint.org/docs/rules/func-call-spacing) | ["error", "never"] | 4 | Yes | Stylistic | |
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab"] | 14604 | Yes | Stylistic | Alt; mostly due to not indenting after the wrapper function in each module|
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0 }] | 4484 | Yes | Stylistic | Alt; allow 0 indent after wrapper function, seems reasonable|
| ✔️  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0, "SwitchCase": 1 }] | 3303 | Yes | Stylistic | Alt; enforce a tab before `case` statements|
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0, "ArrayExpression": "off", "ObjectExpression": "off" }] | 2082 | Yes | Stylistic | Alt; allow Twinkle to use 2 indents instead of 1 in multi-line arrays and objects|
| ❌  |   [indent](https://eslint.org/docs/rules/indent) | ["error", "tab", { "outerIIFEBody": 0, "SwitchCase": 1, "ArrayExpression": "off", "ObjectExpression": "off" }] | 1201 | Yes | Stylistic | Alt; all of the above|
| ❌  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": true, "afterColon": true}}] | 1420 | Yes | Stylistic | Alt|
| ❌  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": true, "afterColon": false}}] | 2476 | Yes | Stylistic | Alt|
| ✔️  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": false, "afterColon": true}}] | 212 | Yes | Stylistic | Alt|
| ❌  |   [key-spacing](https://eslint.org/docs/rules/key-spacing) | ["error", {"singleLine": {"beforeColon": false, "afterColon": false}}] | 1268 | Yes | Stylistic | Alt|
| ✔️  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": true , "before": true}] | 454 | Yes | Stylistic | Alt|
| ❌  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": true , "before": false}] | 734 | Yes | Stylistic | Alt|
| ❌  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": false , "before": true}] | 1333 | Yes | Stylistic | Alt|
| ❌  |   [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing) | ["error", { "after": false , "before": false}] | 1613 | Yes | Stylistic | Alt|
| ✔️  |   [linebreak-style](https://eslint.org/docs/rules/linebreak-style) | ["error", "unix"] | 0 | Yes | Stylistic | |
| ❌  |   [multiline-comment-style](https://eslint.org/docs/rules/multiline-comment-style) | ["error", "separate-lines"] | 24 | Yes | Stylistic |Alt; probably fine to mix-and-match these |
| ❌  |   [multiline-comment-style](https://eslint.org/docs/rules/multiline-comment-style) | ["error", "bare-block"] | 130 | Yes | Stylistic | Alt; probably fine to mix-and-match these|
| ❌  |   [multiline-comment-style](https://eslint.org/docs/rules/multiline-comment-style) | ["error", "starred-block"] | 127 | Yes | Stylistic | Alt; probably fine to mix-and-match these|
| ✔️  |   [no-array-constructor](https://eslint.org/docs/rules/no-array-constructor) | "error" | 0 | No | Stylistic | |
| ✔️  |   [no-bitwise](https://eslint.org/docs/rules/no-bitwise) | "error" | 0 | No |  Stylistic| |
| ❌  |   [no-lonely-if](https://eslint.org/docs/rules/no-lonely-if) | "error" | 6 | Yes (5) | Stylistic | |
| ✔️  |   [no-mixed-operators](https://eslint.org/docs/rules/no-mixed-operators) | "error" | 6 | No | Stylistic | |
| ❔  |   [no-multiple-empty-lines](https://eslint.org/docs/rules/no-multiple-empty-lines) | "error" | 24 | Yes | Stylistic | Default is 2; all but two instances are in morebits, the other are in twinkle.js and xfd |
| ❔  |   [no-multiple-empty-lines](https://eslint.org/docs/rules/no-multiple-empty-lines) | ["error", {"max": 3}] | 0 | Yes | Stylistic | Might be nice for setting apart functions in morebits |
| ✔️  |   [no-new-object](https://eslint.org/docs/rules/no-new-object) | "error" | 0 | No | Stylistic |  | 
| ❌  |   [no-tabs](https://eslint.org/docs/rules/no-tabs) | ["error"] | 17653 | No | Stylistic | Alt; lolno|
| ✔️  |   [no-tabs](https://eslint.org/docs/rules/no-tabs) | ["error", { "allowIndentationTabs": true }] | 11 | No | Stylistic |Alt; will complain about indentation tabs in commented code, though (but we should just remove those) |
| ✔️  |   [no-trailing-spaces](https://eslint.org/docs/rules/no-trailing-spaces) | "error" | 2 | Yes | Stylistic | |
| ✔️  |   [no-unneeded-ternary](https://eslint.org/docs/rules/no-unneeded-ternary) | "error" | 0 | Yes | Stylistic | |
| ✔️  |   [no-whitespace-before-property](https://eslint.org/docs/rules/no-whitespace-before-property) | "error" | 0 | Yes | Stylistic | |
| ❔  |   [operator-linebreak](https://eslint.org/docs/rules/operator-linebreak) | ["error", "after"] | 0 | Yes | Stylistic | |
| ❌  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "always"] | 5219 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "as-needed"] | 500 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "consistent"] | 2 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "consistent-as-needed"] | 393 | Yes | Stylistic | |
| ❔  |   [quote-props](https://eslint.org/docs/rules/quote-props) | ["error", "consistent-as-needed", {"keywords": true}] | 387 | Yes | Stylistic |  |
| ✔️  |   [quotes](https://eslint.org/docs/rules/quotes) | ["error", "single", { "avoidEscape": true }] | 4532 | Yes | Stylistic | Alt; Slight preference (56-44)|
| ❌  |   [quotes](https://eslint.org/docs/rules/quotes) | ["error", "double", { "avoidEscape": true }] | 5790 | Yes | Stylistic | Alt; slight underdog (44-56)|
| ✔️  |   [semi](https://eslint.org/docs/rules/semi) | ["error", "always"] | 2 | Yes | Stylistic | |
| ✔️  |   [space-before-blocks](https://eslint.org/docs/rules/space-before-blocks) | "error" | 60 | Yes | Stylistic | Complains about the current style of the wrapper function for each module|
| ✔️  |   [space-in-parens](https://eslint.org/docs/rules/space-in-parens) | ["error", "never"] | 4452 | Yes | Stylistic | Alt; about time we committed to one style|
| ❌  |   [space-in-parens](https://eslint.org/docs/rules/space-in-parens) | ["error", "always"] | 9340 | Yes | Stylistic | Alt|
| ✔️  |   [space-infix-ops](https://eslint.org/docs/rules/space-infix-ops) | "error" | 67 | Yes | Stylistic | |
| ✔️  |   [space-unary-ops](https://eslint.org/docs/rules/space-unary-ops) | "error" | 20 | Yes | Stylistic | |
| ❌  |   [spaced-comment](https://eslint.org/docs/rules/spaced-comment) | ["error", "never"] | 1168 | Yes | Stylistic | Alt|
| ✔️  |   [spaced-comment](https://eslint.org/docs/rules/spaced-comment) | ["error", "always", { "line": { "exceptions": ["-"] }, "block": { "balanced": true } }] | 86 | Yes | Stylistic | Alt; would require a space before \<nowiki\> tags|
| ❌  |   [spaced-comment](https://eslint.org/docs/rules/spaced-comment) | ["error", "always", { "line": { "exceptions": ["-", "\<nowiki\>"] }, "block": { "balanced": true } }] | 64 | Yes | Stylistic | Alt; don't require a space before \<nowiki\> tags|
| ✔️  |   [switch-colon-spacing](https://eslint.org/docs/rules/switch-colon-spacing) | "error" | 0 | Yes | Stylistic | |

<details>
  <summary>Updated .eslintrc.json with the above options</summary>
  
```json
{
	"extends": ["eslint:recommended"],
	"root": true,
	"env": {
		"browser": true,
		"jquery": true
	},
	"globals": {
		"mw": true,
		"moment": true
	},
	"rules": {
		"no-extra-parens": ["error", "all", { "nestedBinaryExpressions": false }],
		"no-prototype-builtins": "error",
		"block-scoped-var": "error",
		"curly": "error",
		"default-case": "error",
		"eqeqeq": "error",
		"guard-for-in": "error",
		"no-caller": "error",
		"no-else-return": "error",
		"no-implicit-coercion": ["error", { "boolean": false }],
		"no-lone-blocks": "error",
		"no-multi-spaces": ["error", { "ignoreEOLComments": true }],
		"no-useless-catch": "error",
		"no-useless-return": "error",
		"yoda": "error",
		"block-spacing": "error",
		"brace-style": ["error", "1tbs"],
		"comma-dangle": "error",
		"comma-spacing": ["error", { "before": false, "after": true }],
		"computed-property-spacing": ["error", "never"],
		"func-call-spacing": ["error", "never"],
		"indent": ["error", "tab", { "outerIIFEBody": 0, "SwitchCase": 1 }],
		"key-spacing": ["error", {"singleLine": {"beforeColon": false, "afterColon": true}}],
		"keyword-spacing": ["error", { "after": true , "before": true}],
		"linebreak-style": ["error", "unix"],
		"no-array-constructor": "error",
		"no-bitwise": "error",
		"no-mixed-operators": "error",
		"no-new-object": "error",
		"no-tabs": ["error", { "allowIndentationTabs": true }],
		"no-trailing-spaces": "error",
		"no-unneeded-ternary": "error",
		"no-whitespace-before-property": "error",
		"quotes": ["error", "single", { "avoidEscape": true }],
		"semi": ["error", "always"],
		"space-before-blocks": "error",
		"space-in-parens": ["error", "never"],
		"space-infix-ops": "error",
		"space-unary-ops": "error",
		"spaced-comment": ["error", "always", { "line": { "exceptions": ["-"] }, "block": { "balanced": true } }],
		"switch-colon-spacing": "error"
	}
}

```
</details>

### PR DESCRIPTION
**Pull request by Amorymeltzer**
Applies and closes #500; also updates eslint to 5.16.0 (and corresponding dependencies).

----
I've put each new item from #500 in a single commit, so it should be easy to either all at once or piecemeal; I'll squash it when/before merging.  Each commit is labeled with the linting rule being implemented as well as either NOCHANGES, AUTOFIX, or MANUALFIX; generally, the earlier commits are the more stylistic ones while the newer commits are best practices or potential errors.  The manual changes should probably have some more scrutiny than the automatic fixes.  I'll note a few here:

- 2e54a62 `no-mixed-operators`: I think I'm right but it involved parsing out some complex `&&` / `||` stuff
- 0fb3952 `eqeqeq`: Potential for previously-uncaught type issues
- e7a17dd `default-case`: Simple, but I did change the code slightly in one case
- 069d32a `no-extra-parens`: The only case where the autofix introduced an error, in this case `no-cond-assign`.  This is the `while` change on [L1265](https://github.com/azatoth/twinkle/pull/657/commits/069d32a80d0376236280002e99d282f48ea45662#diff-b37f48b65cd39494c40fe24d76458cedR1265) of `twinklewarn.js`

cc maintainers @MusikAnimal @azatoth @kevinji @atlight.  I'd like to keep this open for a week, to give time to review but not too long as to hold up other work by anyone (cc recent contributors @siddharthvp and @DannyS712).  I've pushed this version to testwiki to make it easier to play around with.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -109,11 +109,7 @@ When `deploy`ing or `push`ing, the script will attempt to parse the latest on-wi
 Style guideline
 ---------------
 
-While old legacy code has many different and incoherent styles, it has been decided to utilize a more coherent style throughout the code.
-
-The [jQuery Core Style Guideline][jq_style] is what we will hereafter use as our style guideline.
-
-Needless to say, there are exceptions. The main sticking point is spacing around parentheses. Older Twinkle code looks like `if ( condition ) {`, but newer code tends to use `if (condition) {`. The best convention here is to follow the style of surrounding code.
+While old legacy code previously had many different and incoherent styles, it has been decided to utilize a more coherent style throughout the code, both for consistency and to cut down on potential errors.  [eslint][eslint.org] can be used to check your code before submission - try running `npm run lint` or `eslint .` in your terminal - and can even repair many common issues via `eslint . --fix`.  More information on specific style rules can be seen in [issue #500][fivehundred] and in `.eslintrc.json`, but the best advice is to just follow the style of surrounding code!
```
Comment (MusikAnimal) [inline README.md]: By default eslint wouldn't be installed globally. I think `./node_modules/.bin/eslint` in theory should work for everyone, but the npm scripts are even better. We should add a `fix` script that runs `eslint . --fix`. Also probably worth mentioning [npm](https://www.npmjs.com/) is required (for those who are unfamiliar with it), and that you need to first run `npm install` for the linting scripts to work.

**Comment by MusikAnimal** [comment]
I especially like the spacing rules, we've long needed those!

My feedback:
* https://github.com/azatoth/twinkle/pull/657/commits/65d1aa71a92d8835c1cc6d41acc7a0adf917e770#diff-03744d176dfd1e8a2178545886a87644R644 The old if/else on separate lines, without curly braces, is surely not great. But if we do require curly braces, I think we should enforce if/elses to be multi-line. E.g. `if ( conditional ) { long single-line code } else { long single-line code }` is hard to read, in my opinion. Tag on a few `else if`'s and it's really an eye sore. I shouldn't have to scroll to the right to find out if there's an `else` to go with the `if`.
* https://github.com/azatoth/twinkle/pull/657/commits/f1380b6666c46784ad10385b00663c2b4655cbbb Type-casting using `Boolean` is pretty unusual, from my experience. I think `!!` is fine, but for most of these lines of code, we probably don't need to type-cast anyway.

Otherwise :+1: from me! Code style is of course very subjective. In the end I don't think it matters much, so long as it's consistent.

**Comment by Amorymeltzer** [comment]
@MusikAnimal:
>I think we should enforce if/elses to be multi-line.

I *hugely* agree with you on one-true-brace-style, but given the lack of input at #500 had set it to `allowsingleLine` since there were fewer changes (and I was already feeling like I was imposing).  The end result here is from the combination of [brace-style](https://eslint.org/docs/rules/brace-style) and [curly](https://eslint.org/docs/rules/curly).  I'll **gladly** change the implementation when I update this.

>Type-casting using Boolean is pretty unusual, from my experience.

Again, agreed; I initially had [no-implicit-coercion](https://eslint.org/docs/rules/no-implicit-coercion) set to `["error", { "boolean": false }]` but @kevinji suggested just doing away with them.  We only really use `!!`, and I actually think it's clearer than `Boolean`.

File: README.md (hunk 1 above)
Comment (Amorymeltzer) [reply README.md]: @MusikAnimal Word.  I've got plans to write up and do a lot of that as part of #477 and beyond — including some file reorganization and a /dev folder — but I'll fix up this note here per your recommendations.

**Comment by MusikAnimal** [comment]
> We only really use `!!`, and I actually think it's clearer than `Boolean`

Yes, cleaner, and it carries over to C/C++, PHP, Ruby, etc. I've never seen the JavaScript `Boolean()` in the wild. It isn't wrong, it just looks weird :)

**Comment by kevinji** [comment]
Yeah, on second thought I agree with y'all about the `Boolean`, let's change that back to using `!!`.

**Comment by siddharthvp** [comment]
For the unused index reference in `.each()` loops, some code editors display a warning and recommend prefixing the variable name with underscore, that is using `_i` instead of `i`. This practise seems to be followed in many JS projects I've seen. Should this be included in the ruleset? Looks like this can be enforced using `no-unused-vars: ["error", { "vars": "all", "args": "all", "argsIgnorePattern": "^_" }]`, per https://eslint.org/docs/rules/no-unused-vars.

**Comment by siddharthvp** [comment]
https://github.com/azatoth/twinkle/commit/65d1aa71a92d8835c1cc6d41acc7a0adf917e770#diff-03744d176dfd1e8a2178545886a87644L1878, this is no good. We should allow `if` without curly braces. It isn't useful on most occasions, but it's ok when:
- The single-line code is just `return` or `continue`.
- There is a huge list of `if` conditions with single-line code, then it is concise to omit the braces and multi-lining.

**Comment by siddharthvp** [comment]
https://github.com/azatoth/twinkle/pull/657/commits/e8b508424cb7f1def0d9b2df5308ceec3e743524 I prefer writing `if (! condition)` since`if (!condition)` is easily misread as `if (condition)`. `if ( !condition )` is ok too, though conflicting with the rule on no spacing.

**Comment by siddharthvp** [comment]
https://github.com/azatoth/twinkle/pull/657/commits/30a5b76f38bd41392a2e1634436a40492775965b empty functions shouldn't be multiline - that's just a waste of 2 lines. And the `// do nothing` comment isn't needed in `emptyFunction` coz the name of the function makes it clear its gonna do nothing. The linter rule is not worth adding because empty functions are so infrequent.

**Comment by siddharthvp** [comment]
https://github.com/azatoth/twinkle/pull/657/commits/ff8d55baf6a1520c895f409b78058b55f18b9ef7 I think `if (condition) return x;  else return y;` is more reader-friendly than `if (condition) return x;  return y;`  Though when the else statement is not a return statement, the latter style would be better.

**Comment by siddharthvp** [comment]
https://github.com/azatoth/twinkle/pull/657/commits/e7a17dd9a9f3fbae13bcb517d7f6bd4fe07bc78c#diff-03744d176dfd1e8a2178545886a87644L418 Don't see any point. This just wastes 4 lines and I think the earlier version is more readable.

**Comment by siddharthvp** [comment]
- https://github.com/azatoth/twinkle/commit/65d1aa71a92d8835c1cc6d41acc7a0adf917e770#diff-03744d176dfd1e8a2178545886a87644L1878, this is no good. We should allow `if` without curly braces. It isn't useful on most occasions, but it's ok when:
  - The single-line code is just `return` or `continue`.
  - There is a huge list of `if` conditions with single-line code, then it is concise to omit the braces and multi-lining. 

- https://github.com/azatoth/twinkle/pull/657/commits/e8b508424cb7f1def0d9b2df5308ceec3e743524 I prefer writing `if (! condition)` since`if (!condition)` is easily misread as `if (condition)`. `if ( !condition )` is ok too, though conflicting with the rule on no spacing.

- https://github.com/azatoth/twinkle/pull/657/commits/30a5b76f38bd41392a2e1634436a40492775965b empty functions shouldn't be multiline - that's just a waste of 2 lines. And the `// do nothing` comment isn't needed in `emptyFunction` coz the name of the function makes it clear its gonna do nothing. The linter rule is not worth adding because empty functions are so infrequent.

- https://github.com/azatoth/twinkle/pull/657/commits/ff8d55baf6a1520c895f409b78058b55f18b9ef7 I think `if (condition) return x;  else return y;` is more reader-friendly than `if (condition) return x;  return y;`  Though when the else statement is not a return statement, the latter style would be better.

- https://github.com/azatoth/twinkle/pull/657/commits/e7a17dd9a9f3fbae13bcb517d7f6bd4fe07bc78c#diff-03744d176dfd1e8a2178545886a87644L418 Don't see any point. This just wastes 4 lines and I think the earlier version is more readable.

- https://github.com/azatoth/twinkle/pull/657/commits/372e2306e5669aa83b0837b51b8e63e169a4533d What's wrong with not respecting block scope? The particular usage looked better before the "fix".

- https://github.com/azatoth/twinkle/pull/657/commits/069d32a80d0376236280002e99d282f48ea45662 Maybe a net positive, but doesn't it make some ternaries unreadable? Parens are supposed to help computers to parse but sometimes they help humans too.

- For the unused index reference in `.each()` loops, some code editors display a warning and recommend prefixing the variable name with underscore, that is using `_i` instead of `i`. This practise seems to be followed in many JS projects I've seen. Should this be included in the ruleset? The rule will be `no-unused-vars: ["error", { "vars": "all", "args": "all", "argsIgnorePattern": "^_" }]`, per https://eslint.org/docs/rules/no-unused-vars.

The spacing changes look really good, though 👍.

**Comment by Amorymeltzer** [comment]
@siddharthvp:

>We should allow if without curly braces

Per above convo with @MusikAnimal, these will all become multi-line, but in theory I agree; I think the strict enforcement of `curly` was held over from before I felt guilty about unilaterally imposing 1tbs.

>I prefer writing if `(! condition)`

This would be `"space-unary-ops": ["error", {"words": true, "nonwords": false, "overrides": {"!": true}}]` I think, and as of this writing would require 294 changes as opposed to just 25 with the default.  That's a pretty clear house style I think.

>The linter rule is not worth adding because empty functions are so infrequent

I'm fine with that, although I suppose my concern was less about readability for simple functions like these and rather some of the more esoteric ones.  I think we could also do `"allow": ["functions"]`, but that's even less likely to get used.

>I think `if (condition) return x; else return y;` is more reader-friendly than `if (condition) return x; return y;`

I don't disagree, but I think that's because I'm used to thinking in an if-else paradigm, where there's code after, something like
```javascript
if (condition) {
	x = x + 4;
} else {
	x = x + 2;
}
alert(x);
```
But when `return`ing in all cases, there's no code that will come after, so I'm not really right to think of it as an if-else rather than just `if`, somewhat similar to having a default `case` (see below) to fall back to.  I don't like this rule, but I think it makes reading it more correct.

>This just wastes 4 lines and I think the earlier version is more readable

Fair point, will revert.  In hindsight, I probably shouldn't have anyway, it's not linting.

>What's wrong with not respecting block scope?

Best practices.  js is nicer with variables than some other languages, but it's clearer that these are expected to be used elsewhere.  We're helped here by the very nature of the form, but declaring a variable within an `if` and using it elsewhere is a good way to be confusing.  Also, before that recent commit, we'd done it everywhere else.

>doesn't it make some ternaries unreadable? Parens are supposed to help computers to parse but sometimes they help humans too.

We were real bad with consistency on this front, with ternary parens used all over the place for readability.  Again, I don't really disagree, but I think having `((x===y) ? a: b)` and `x===y ? a: b` and `(x===y ? a: b)` and `(x===y) ? a: b` is likely to breed more confusion.

> some code editors display a warning and recommend prefixing the variable name with underscore, that is using `_i` instead of `i`

You fancy people with your fancy editors!  I'm not sure what you're asking, though; don't we already cover this with `"extends": "eslint:recommended"`?  Or does your editor *require* an unused `_i` variable?

**Comment by siddharthvp** [comment]
> Per above convo with @MusikAnimal, these will all become multi-line,

I think multiline-with-braces and no-multiline-no-braces should be allowed. Other combinations should be disallowed.

>  I think, and as of this writing would require 294 changes as opposed to just 25 with the default. 

I meant more like not disallowing `if (! condition)` rather than enforcing it.

> don't we already cover this with "extends": "eslint:recommended"? Or does your editor require an unused _i variable?

I'm not familiar with eslint, so not sure, but I guess the default would be `no-unused-vars: ["error", { "vars": "all", "args": "after-used" } ]`. It doesn't require it of course, but it shows the greys out the `i` and hovering shows `'i' is declared but its value is never read.ts(6133)` along with an option to automatically convert it to `_i`. This is VS code (use it, it's the best).

**Comment by MusikAnimal** [comment]
>> Per above convo with @MusikAnimal, these will all become multi-line,
>
> I think multiline-with-braces and no-multiline-no-braces should be allowed. Other combinations should be disallowed.

This was my thinking, I just didn't say it out loud. So, good:
```
if (foo) {
  alert('bar');
} else {
  alert('baz');
}
```
```
if (foo) {
  bar += baz + 5;
}
```
Bad:
```
if (foo) alert('bar'); else alert('baz');
```
```
if (foo) { bar += baz + 5; }
```
Really bad:
```
if (foo) bar += baz + 5;
else bar -= baz;
``` 
Okay:
```
if (foo) return;
```
Not so okay:
```
if (foo) alert('bar');
```

Basically what @siddharthvp said; `if` without curly braces is fine if it's solely a terminating statement such as `return` or `continue`. Even `if (foo) return baz + 5;` is not great, in my opinion. A long list of single-line `if`s may look more concise without curlys, but that might put into question whether there should be a long list of conditionals. We're light years away from achieving it, but in general you want to keep a given function's [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) to a minimum.

I've no idea how easy or feasible it is to enforce the above rules with eslint, just sharing the best practices I'm used to.

----
I am also familiar with prefixing unused vars with `_` and agree with this practice. In general I don't think we should pay much mind to the preferences of code editors, though; they should respect .eslint.json and nothing else (your editor is probably just indicating unused vars for convenience, as opposed to suggesting they are true errors).

**Comment by siddharthvp** [comment]
> We're light years away from achieving it, but in general you want to keep a given function's cyclomatic complexity to a minimum.

You're right of course, it's quite possible to rewrite the code using which Tag constructs the template text, removing the large switch and list of the conditionals in favour of an object-oriented approach, the way I've done it for talk tags in #488.

**Comment by Amorymeltzer** [comment]
I've been meaning to try out VS code, mostly just because everyone raves, but even emacs with `js2-mode` and flymake (not even flycheck!) gives me warnings/alerts that eslint doesn't.  I do think unused vars are covered by `"extends": "eslint:recommended"`.

And AFAIK eslint doesn't allow specifying when a single line `if` would be okay.  I agree with both of ya, for sure, but won't let us pick and choose a few exceptions; it's enforce multiline `if`-`else` or not.

Not necessarily related to the above, but speaking more generally, I think it's good to be as clear and open as possible in our code.  While this is technically a single person's github repo to which a small number of thoughtful and committed (and kind and smart and attractive and generous and...) people contribute, I think it's also good to be cognizant of this being a wiki[pm]edia project, so doing what we can to make it easier for folks to read, understand, and contribute (while not making our lives miserable) is a noble goal.

File: README.md (hunk 1 above)
Comment (Amorymeltzer) [reply README.md]: @MusikAnimal Rewritten, will expand as part of 477.  I'll probably add a `lintfix` (`fixlint`?) shortcut then as well, but given how we define `lint` we can just run `npm run lint -- --fix` to fix things at the moment.

**Comment by Amorymeltzer** [comment]
Force-pushed update, incorporates some of the discussion above and rebased.

File: README.md (hunk 1 above)
Comment (MusikAnimal) [reply README.md]: :+1:

**Comment by Amorymeltzer** [comment]
Merged, thanks everybody.  I'm gonna push this live in the interest of making future on-wiki diffs easier to parse, and to catch any issues sooner rather than later.

**Pull request merged**